src/rdb_protocol/terms/string.cc

```diff
@@ -125,6 +127,64 @@ class match_term_t : public op_term_t {
     virtual const char *name() const { return "match"; }
 };
 
+class format_term_t : public op_term_t {
+public:
+    format_term_t(compile_env_t *env, const raw_term_t &term)
+            : op_term_t{env, term, argspec_t{2}}, src_term{term}, compile_env{env} {}
+private:
+    virtual scoped_ptr_t <val_t> eval_impl(scope_env_t *env, args_t *args, eval_flags_t) const {
+        const std::string templ = args->arg(env, 0)->as_str().to_std();
+        const datum_t datum = args->arg(env, 1)->as_datum();
+
+        const size_t template_length = templ.size();
+        std::string formatted{};
+
+        for (size_t pos = 0; pos < template_length; ++pos) {
+            rcheck(templ[pos] != '}',
+                   base_exc_t::LOGIC,
+                   "No parameter tags to close");
+
+            if (templ[pos] == '{') {
+                size_t param_end_pos {1};
+                std::string param_name {};
+
+                while (pos + param_end_pos < template_length && templ[pos + param_end_pos] != '}') {
+                    char next_char{templ[pos + param_end_pos]};
+
+                    rcheck(next_char != '{',
+                           base_exc_t::LOGIC,
+                           "Nested template parameters are not allowed");
+
+                    param_name.push_back(next_char);
+                    ++param_end_pos;
+                }
+
+                rcheck(pos + param_end_pos != template_length,
+                       base_exc_t::LOGIC,
+                       "Parameter tag must be closed");
+
+                minidriver_t r{src_term.bt()};
+                datum_t field{datum.get_field(datum_string_t{param_name})};
```

Comment:
A question here, also with coerceTo('STRING'), is what behavior we want if both the format string has parse errors, and some early fields are missing.  Do we want to parse the whole format string first, and produce an error that there is an invalid format string?  Or is having errors on field lookups or coerceTo('STRING') behavior (if that can error) desirable?

Reply:
I think throwing an error on the case whichever happens first is desirable and expected.

If someone wants to format the content of _War and Peace_ from Tolstoy in a way that it adds a pre- and postfix to the book's content, and the first parameter is missing, we don't want to parse the whole text if we already know there is an error. So we raise an error at the time we realize a parameter is missing (, obviously, we cannot do much about the case if the suffix is missing). My thinking is the same for `coerceTo`, if we know we cannot coerce the field for some reason, we should let the caller know it as soon as possible.

Reply:
If the format string has parse errors, we would be saving work by not coercing some of the parameters before finding the parse error.

But I'm not concerned about the performance here.  What I'm concerned about is providing a useful error message to the user, and avoiding behavior changes in the future.

Reply:
`to_std()` is what creates the copy here.  What I meant was something like

```
datum_string_t result = term->eval(env)->as_str();  // Or use ->as_datum(), then use a const datum_string_t& variable?
formatted.append(result.data(), result.size());
```

Note that `val_t` says to prefer `val_t::as_str()` with its comment, "prefer the forms below".  It's unclear to me why, but it catches the `datum_exc_t` and turns it into an `rfail`.

Reply:
@srh The copying issue is resolved by [45d0277](https://github.com/rethinkdb/rethinkdb/pull/7066/commits/45d02778807a2036b46433c0ced945e689d2d31a). Regarding the formatting and https://github.com/rethinkdb/rethinkdb/pull/7066#issuecomment-1133498128, what would you suggests?

In terms of keeping characters for formatting, I would personally prefer the colon approach or percent sign. These two characters are rarely used in document keys -- I hope -- and both are used by other languages like Python (`%`), Rust (`:`), Java (`%`), etc.

As of error messages, what are you suggesting? Should we parse the whole template string before trying to look for parameter replacements?

Reply:
Let me talk only about the error behavior issue here.

My question is, which error behavior is better?  Does it even matter?  The implementation might be fine.

Reply:
> which error behavior is better? Does it even matter? 

Honestly speaking, I don't know which behaviour would be better. I think this is something that can be changed later on if the community prefers the other way around that we implement. I would leave it as it is now and change if someone has serious issues with this approach.